// @lwrjs/app-service/build/cjs/moduleProvider/index.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __exportStar = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  return __exportStar(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? {get: () => module2.default, enumerable: true} : {value: module2, enumerable: true})), module2);
};

// packages/@lwrjs/app-service/src/moduleProvider/index.ts
__markAsModule(exports);
__export(exports, {
  default: () => moduleProvider_default
});
var import_diagnostics = __toModule(require("@lwrjs/diagnostics"));
var import_shared_utils = __toModule(require("@lwrjs/shared-utils"));
var import_utils = __toModule(require("./utils.cjs"));
var import_identity = __toModule(require("../identity.cjs"));
function validateSpecifier(specifier) {
  const appIdentity = (0, import_identity.parseAppIdentity)(specifier);
  return !!appIdentity && (0, import_identity.isValidResourceIdentity)(appIdentity, import_identity.AppResourceEnum.MODULE);
}
var AppModuleProvider = class {
  constructor(appPluginConfig, {config, runtimeEnvironment: {hmrEnabled, lwrVersion}}) {
    this.name = "app-provider";
    this.hmrEnabled = false;
    this.config = config;
    this.hmrEnabled = hmrEnabled;
    this.version = lwrVersion;
    this.lockerConfig = config.locker;
  }
  async getModuleEntry({specifier}) {
    if (validateSpecifier(specifier)) {
      const normalizedSpecifier = specifier.endsWith("/") ? `${specifier}index` : specifier;
      const virtualId = `<virtual>/${normalizedSpecifier}`;
      return {
        id: `${virtualId}|${this.version}`,
        virtual: true,
        entry: virtualId,
        specifier,
        version: this.version
      };
    }
  }
  async getModuleSource({
    specifier,
    namespace,
    name
  }) {
    const appIdentity = (0, import_identity.parseAppIdentity)(specifier);
    if (!appIdentity || !(0, import_identity.isValidResourceIdentity)(appIdentity, import_identity.AppResourceEnum.MODULE)) {
      return;
    }
    const moduleEntry = await this.getModuleEntry({specifier});
    const route = [...this.config.routes, ...this.config.errorRoutes].find((r) => r.id === appIdentity.appName);
    if (!moduleEntry || !route) {
      throw (0, import_diagnostics.createSingleDiagnosticError)({
        description: import_diagnostics.descriptions.UNRESOLVABLE.MODULE(specifier)
      }, import_diagnostics.LwrUnresolvableError);
    }
    const {format} = appIdentity;
    const moduleLoader = format === "amd" ? this.config.amdLoader : this.config.esmLoader;
    let originalSource;
    if (appIdentity.subResource && (0, import_identity.isWorkerSubResource)(appIdentity.subResource)) {
      const workerId = appIdentity.subResource.workerId;
      const workerSpecifier = route.bootstrap.workers[workerId];
      if (!workerSpecifier) {
        throw (0, import_diagnostics.createSingleDiagnosticError)({
          description: import_diagnostics.descriptions.UNRESOLVABLE.MODULE(specifier)
        }, import_diagnostics.LwrUnresolvableError);
      }
      originalSource = (0, import_utils.createAppWorkerBootstrapModule)(workerSpecifier, route, {
        format,
        moduleLoader,
        hmrEnabled: false
      });
    } else {
      originalSource = (0, import_utils.createAppRouteViewBootstrapModule)(route, {
        format,
        hmrEnabled: this.hmrEnabled,
        moduleLoader
      }, this.lockerConfig);
    }
    return {
      id: moduleEntry.id,
      namespace,
      name: name || specifier,
      version: this.version,
      specifier,
      moduleEntry,
      ownHash: (0, import_shared_utils.hashContent)(originalSource),
      originalSource
    };
  }
  async getModule(moduleId) {
    if (!validateSpecifier(moduleId.specifier)) {
      return;
    }
    const moduleSource = await this.getModuleSource(moduleId);
    const compiledSource = moduleSource.originalSource;
    return {
      ...moduleSource,
      compiledSource
    };
  }
};
var moduleProvider_default = AppModuleProvider;
