// @lwrjs/shared-utils/build/es/import-metadata.js
import { parse as parseImports } from 'es-module-lexer';
export var ModuleNameType;
(function (ModuleNameType) {
    ModuleNameType["string"] = "string";
    ModuleNameType["unresolved"] = "unresolved";
})(ModuleNameType || (ModuleNameType = {}));
export async function getImportMetadata(compiledSource) {
    const imports = [];
    const dynamicImports = [];
    // Use very fast parser to get the location of the imports/exports
    const [moduleImportLocations] = await parseImports(compiledSource);
    for (const moduleImportLocation of moduleImportLocations) {
        // This will give us the source specifier of the import (ex. import x from "bar" <= will get us bar)
        let moduleSpecifier = compiledSource.substring(moduleImportLocation.s, moduleImportLocation.e);
        const isDynamic = moduleImportLocation.d !== -1;
        const location = {
            startColumn: moduleImportLocation.s,
            endColumn: moduleImportLocation.e,
        };
        if (isDynamic) {
            const isStringLiteral = moduleSpecifier.startsWith("'") || moduleSpecifier.startsWith('"');
            const moduleNameType = isStringLiteral ? ModuleNameType.string : ModuleNameType.unresolved;
            moduleSpecifier = isStringLiteral ? moduleSpecifier.slice(1, -1) : moduleSpecifier; // remove quotes
            dynamicImports.push({
                moduleSpecifier,
                moduleNameType,
                location,
                importLocation: {
                    startColumn: moduleImportLocation.d,
                    endColumn: moduleImportLocation.s,
                },
            });
        }
        else {
            imports.push({ moduleSpecifier, location });
        }
    }
    return {
        imports,
        dynamicImports,
    };
}
