// @lwrjs/nunjucks-view-provider/build/cjs/index.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __exportStar = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  return __exportStar(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? {get: () => module2.default, enumerable: true} : {value: module2, enumerable: true})), module2);
};

// packages/@lwrjs/nunjucks-view-provider/src/index.ts
__markAsModule(exports);
__export(exports, {
  default: () => src_default
});
var import_path = __toModule(require("path"));
var import_nunjucks = __toModule(require("nunjucks"));
var import_gray_matter = __toModule(require("gray-matter"));
var import_shared_utils = __toModule(require("@lwrjs/shared-utils"));
var import_nunjucks_env_utils = __toModule(require("./nunjucks-env-utils.cjs"));
var NunJucksViewProvider = class {
  constructor(pluginConfig, providerConfig) {
    this.watcherViewMap = new Map();
    this.routes = providerConfig.config.routes;
    this.contentDir = providerConfig.config.contentDir;
    this.layoutsDir = providerConfig.config.layoutsDir;
    this.rootDir = providerConfig.config.rootDir;
    this.name = "lwr-nunjucks-view-provider";
    this.emitter = providerConfig.appEmitter;
    const {
      runtimeEnvironment: {watchFiles}
    } = providerConfig;
    if (watchFiles) {
      this.watcher = (0, import_shared_utils.setupWatcher)(this.onViewChange.bind(this));
    }
  }
  async onViewChange(fileChanged) {
    const cachedView = this.watcherViewMap.get(fileChanged);
    if (!cachedView) {
      throw new Error("We are observing a file we have not yet processed, this shouldnt happen...");
    }
    const compiledView = await this.getView(cachedView.viewId);
    if (!compiledView) {
      throw new Error("Unable to get compiledView");
    }
    this.emitter.notifyViewSourceChanged(compiledView);
  }
  async initialize() {
    return;
  }
  watchView(fullPath, viewId) {
    if (this.watcher && !this.watcherViewMap.has(fullPath)) {
      this.watcherViewMap.set(fullPath, {viewId});
      this.watcher.add(fullPath);
    }
  }
  async getView(viewId) {
    if (viewId.rootComponent) {
      return;
    }
    const sourcePath = viewId.contentTemplate;
    if (!sourcePath || !(0, import_shared_utils.canResolveView)(sourcePath, "njk")) {
      return;
    }
    const sourceDir = import_path.default.dirname(sourcePath);
    const viewSource = (0, import_shared_utils.getViewSourceFromFile)(sourcePath);
    const {content, data} = (0, import_gray_matter.default)(viewSource.originalSource);
    const {contentDir, layoutsDir, rootDir, watcher} = this;
    const nunjucksCustomLoader = new import_nunjucks_env_utils.NunjucksCustomLoader({contentDir, layoutsDir, rootDir, sourceDir});
    if (watcher) {
      nunjucksCustomLoader.on("import", (importFile) => this.watchView(importFile, viewId));
    }
    const nunjucksEnv = new import_nunjucks.default.Environment(nunjucksCustomLoader);
    (0, import_nunjucks_env_utils.addHelperMethods)(nunjucksEnv);
    const nunjucksView = import_nunjucks.default.compile(content, nunjucksEnv);
    const {layoutTemplate, immutable, ...nonLwrManagedMetadata} = data;
    const compiledView = {
      ...viewSource,
      viewId,
      layoutTemplate,
      immutable,
      properties: {
        ...nonLwrManagedMetadata
      },
      render: async (context) => {
        return {
          renderedView: nunjucksView.render(context || {})
        };
      }
    };
    const fullPath = import_path.default.resolve(sourcePath);
    if (watcher) {
      this.watchView(fullPath, viewId);
    }
    return compiledView;
  }
};
var src_default = NunJucksViewProvider;
