// @lwrjs/server/build/cjs/fs/fs-request.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __exportStar = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  return __exportStar(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? {get: () => module2.default, enumerable: true} : {value: module2, enumerable: true})), module2);
};

// packages/@lwrjs/server/src/fs/fs-request.ts
__markAsModule(exports);
__export(exports, {
  default: () => fs_request_default
});
var import_koa_request = __toModule(require("../koa/koa-request.cjs"));
var FsRequest = class extends import_koa_request.default {
  get query() {
    return this.ctx.request.query || "";
  }
  acceptsLanguages() {
    return [this.ctx.locale];
  }
  hasJsonParam() {
    return this.ctx.isJsonRequest;
  }
  isJsonRequest() {
    return this.ctx.isJsonRequest;
  }
  validateJsonRequest() {
    return true;
  }
  isSiteGeneration() {
    return true;
  }
};
var fs_request_default = FsRequest;
