// @lwrjs/server/build/cjs/fs/fs-response.cjs
var __defProp = Object.defineProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};

// packages/@lwrjs/server/src/fs/fs-response.ts
__markAsModule(exports);
__export(exports, {
  default: () => fs_response_default
});
var FsResponse = class {
  constructor(context) {
    if (!context.fs) {
      context.fs = {headers: {}};
    }
    this.context = context;
  }
  set(headers) {
    this.context.fs.headers = {...this.context.headers, ...headers};
  }
  setHeader(name, value) {
    this.context.fs.headers[name] = value;
  }
  status(status) {
    this.context.status = status;
    return this;
  }
  send(body) {
    this.context.fs.body = body;
  }
  setMetadata(metadata) {
    this.context.fs.metadata = metadata;
  }
  sendStatus(httpCode) {
    return;
  }
  stream(stream) {
    this.context.fs.stream = stream;
  }
  type(type) {
    this.context.fs.type = type;
    return this;
  }
  cookie(name, value, options) {
    return this;
  }
};
var fs_response_default = FsResponse;
