// @lwrjs/server/build/cjs/fs/fs-server.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __exportStar = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  return __exportStar(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? {get: () => module2.default, enumerable: true} : {value: module2, enumerable: true})), module2);
};

// packages/@lwrjs/server/src/fs/fs-server.ts
__markAsModule(exports);
__export(exports, {
  FsInternalServer: () => FsInternalServer,
  LwrFsKoa: () => LwrFsKoa
});
var import_koa = __toModule(require("koa"));
var import_router = __toModule(require("@koa/router"));
var import_koa_compose = __toModule(require("koa-compose"));
var import_http = __toModule(require("http"));
var import_fs_request = __toModule(require("./fs-request.cjs"));
var import_fs_response = __toModule(require("./fs-response.cjs"));
var import_fs_context = __toModule(require("./fs-context.cjs"));
var import_abstract_internal_app_server = __toModule(require("../common/abstract-internal-app-server.cjs"));
var LwrFsKoa = class extends import_koa.default {
  constructor() {
    super();
  }
  createMockContext(url, method = "GET", lang) {
    return (0, import_fs_context.default)(this, url, method, lang);
  }
  setMiddlewares(middlewares) {
    this.composedMiddleware = (0, import_koa_compose.default)(middlewares);
  }
  async dispatchUrl(url, method, lang) {
    const {composedMiddleware} = this;
    if (!composedMiddleware) {
      throw new Error(`Middlewares not initialized`);
    }
    const ctx = this.createMockContext(url, method, lang);
    await composedMiddleware(ctx);
    return ctx;
  }
};
var FsInternalServer = class extends import_abstract_internal_app_server.default {
  constructor() {
    super();
    this.middlewares = [];
    this.lwrFsKoaApp = new LwrFsKoa();
    this.koaRouter = new import_router.default();
  }
  createHttpServer() {
    return import_http.default.createServer();
  }
  use(middleware) {
    this.middlewares.push(async (ctx, next) => {
      const lwrReq = new import_fs_request.default(ctx);
      const lwrRes = new import_fs_response.default(ctx);
      await middleware(lwrReq, lwrRes, next);
    });
  }
  all(path, middleware) {
    this.koaRouter.all(path, async (ctx, next) => {
      const lwrReq = new import_fs_request.default(ctx);
      const lwrRes = new import_fs_response.default(ctx);
      await middleware(lwrReq, lwrRes, next);
    });
  }
  get(path, middleware) {
    this.koaRouter.get(path, async (ctx, next) => {
      const lwrReq = new import_fs_request.default(ctx);
      const lwrRes = new import_fs_response.default(ctx);
      await middleware(lwrReq, lwrRes, next);
    });
  }
  post(path, middleware) {
    this.koaRouter.post(path, async (ctx, next) => {
      const lwrReq = new import_fs_request.default(ctx);
      const lwrRes = new import_fs_response.default(ctx);
      await middleware(lwrReq, lwrRes, next);
    });
  }
  initRoutes() {
    this.middlewares.push(this.koaRouter.routes());
    this.lwrFsKoaApp.setMiddlewares(this.middlewares);
  }
  useCompression() {
  }
  getImpl() {
    return this.lwrFsKoaApp;
  }
  getRegexWildcard() {
    return "(.*)";
  }
};
