// @lwrjs/server/build/es/fs/fs-response.js
export default class FsResponse {
    constructor(context) {
        if (!context.fs) {
            context.fs = { headers: {} };
        }
        this.context = context;
    }
    set(headers) {
        this.context.fs.headers = { ...this.context.headers, ...headers };
    }
    setHeader(name, value) {
        this.context.fs.headers[name] = value;
    }
    status(status) {
        this.context.status = status;
        return this;
    }
    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    send(body) {
        this.context.fs.body = body;
    }
    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    setMetadata(metadata) {
        this.context.fs.metadata = metadata;
    }
    sendStatus(httpCode) {
        return;
    }
    stream(stream) {
        this.context.fs.stream = stream;
    }
    type(type) {
        this.context.fs.type = type;
        return this;
    }
    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    cookie(name, value, options) {
        return this;
    }
}
