// @lwrjs/shared-utils/build/cjs/fs.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __exportStar = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  return __exportStar(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? {get: () => module2.default, enumerable: true} : {value: module2, enumerable: true})), module2);
};

// packages/@lwrjs/shared-utils/src/fs.ts
__markAsModule(exports);
__export(exports, {
  canResolveView: () => canResolveView,
  createFileWatcher: () => createFileWatcher,
  getViewSourceFromFile: () => getViewSourceFromFile,
  hashContent: () => hashContent,
  mimeLookup: () => import_mime_types.lookup,
  normalizeDirectory: () => normalizeDirectory,
  normalizeResourcePath: () => normalizeResourcePath,
  readFile: () => readFile,
  resolveFileExtension: () => resolveFileExtension,
  setupWatcher: () => setupWatcher
});
var import_fs = __toModule(require("fs"));
var import_path = __toModule(require("path"));
var import_crypto = __toModule(require("crypto"));
var import_identity = __toModule(require("./identity.cjs"));
var import_object = __toModule(require("./object.cjs"));
var import_chokidar = __toModule(require("chokidar"));
var import_diagnostics = __toModule(require("@lwrjs/diagnostics"));
var import_mime_types = __toModule(require("mime-types"));
var import_logger = __toModule(require("./logger.cjs"));
function hashContent(source) {
  return import_crypto.default.createHash("md5").update(source).digest("hex");
}
function readFile(filePath) {
  return import_fs.default.readFileSync(filePath, "utf8");
}
function resolveFileExtension(filePath) {
  const fileName = import_path.default.basename(filePath);
  const hasExt = !!import_path.default.extname(fileName);
  if (hasExt) {
    return filePath;
  }
  if (import_fs.default.existsSync(filePath)) {
    if (import_fs.default.statSync(filePath).isFile()) {
      return filePath;
    } else {
      filePath = import_path.default.join(filePath, "index");
    }
  }
  if (import_fs.default.existsSync(filePath + ".js")) {
    return filePath + ".js";
  } else if (import_fs.default.existsSync(filePath + ".ts")) {
    return filePath + ".ts";
  } else {
    throw new Error(`Unable to find file "${filePath}"`);
  }
}
function createFileWatcher(options = {persistent: true, ignored: "**/node_modules/**"}) {
  return import_chokidar.default.watch([], options);
}
function setupWatcher(onModuleChange) {
  const fileWatcher = createFileWatcher();
  fileWatcher.on("change", (0, import_object.debounce)((file) => onModuleChange(file), 500));
  fileWatcher.on("unlink", (0, import_object.debounce)((file) => onModuleChange(file), 500));
  fileWatcher.on("add", (file) => import_logger.logger.info(`Watching: ${file}`));
  return fileWatcher;
}
function canResolveView(source, type) {
  if (!source.endsWith(`.${type}`)) {
    return false;
  }
  if (!import_fs.default.existsSync(import_path.default.resolve(source))) {
    throw new Error(`View template cannot be found: "${source}"`);
  }
  return true;
}
function getViewSourceFromFile(source) {
  const sourceFilePath = import_path.default.resolve(source);
  const viewSource = readFile(sourceFilePath);
  const ext = import_path.default.extname(sourceFilePath);
  const name = import_path.default.basename(sourceFilePath, ext);
  return {
    name,
    slug: (0, import_identity.slugify)(name),
    filePath: sourceFilePath,
    ownHash: hashContent(viewSource),
    originalSource: viewSource
  };
}
var ROOT_DIR_REGEX = /\$rootDir/g;
function normalizeDirectory(dir, rootDir) {
  return dir.replace(ROOT_DIR_REGEX, rootDir);
}
var RESOURCE_DIR_REGEX = /\$(\w+)|^\$(\w+)/g;
function normalizeResourcePath(rawPath, {rootDir, assets, contentDir, layoutsDir}, allowUnresolvedAlias) {
  const assetsMap = assets.reduce((map, asset) => {
    if (asset.alias) {
      map[asset.alias] = asset.dir || asset.file;
    }
    return map;
  }, {rootDir, contentDir, layoutsDir});
  return rawPath.replace(RESOURCE_DIR_REGEX, function(fullMatch, m1, m2) {
    const alias = assetsMap[m1] || assetsMap[m2];
    if (!alias && !allowUnresolvedAlias) {
      throw (0, import_diagnostics.createSingleDiagnosticError)({
        description: import_diagnostics.descriptions.UNRESOLVABLE.DIR_ALIAS(fullMatch)
      }, import_diagnostics.LwrUnresolvableError);
    }
    return alias;
  });
}
