// @lwrjs/router/build/cjs/modules/lwr/navigation/navigation.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __exportStar = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  return __exportStar(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? {get: () => module2.default, enumerable: true} : {value: module2, enumerable: true})), module2);
};

// packages/@lwrjs/router/src/modules/lwr/navigation/navigation.ts
__markAsModule(exports);
__export(exports, {
  ContextInfo: () => import_contextUtils.ContextInfo,
  CurrentPageReference: () => import_currentPageReference.CurrentPageReference,
  CurrentView: () => import_currentView.CurrentView,
  NavigationContext: () => import_navigationContext.NavigationContext,
  NavigationMixin: () => import_navigationMixin.NavigationMixin,
  generateContextualWireAdapter: () => import_contextUtils.generateContextualWireAdapter,
  generateUrl: () => import_navigationApi.generateUrl,
  getNavigationHelm: () => import_contextUtils.getNavigationHelm,
  navigate: () => import_navigationApi.navigate,
  registerNavigationHelm: () => import_contextUtils.registerNavigationHelm
});
var import_navigationContext = __toModule(require("lwr/navigationContext"));
var import_currentPageReference = __toModule(require("lwr/currentPageReference"));
var import_currentView = __toModule(require("lwr/currentView"));
var import_contextUtils = __toModule(require("lwr/contextUtils"));
var import_navigationApi = __toModule(require("./navigationApi"));
var import_navigationMixin = __toModule(require("./navigationMixin"));
