// @lwrjs/lwc-ssr/build/cjs/moduleProvider/index.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __exportStar = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  return __exportStar(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? {get: () => module2.default, enumerable: true} : {value: module2, enumerable: true})), module2);
};

// packages/@lwrjs/lwc-ssr/src/moduleProvider/index.ts
__markAsModule(exports);
__export(exports, {
  createSsrBootstrapModule: () => createSsrBootstrapModule,
  default: () => moduleProvider_default
});
var import_shared_utils = __toModule(require("@lwrjs/shared-utils"));
var import_identity = __toModule(require("../identity.cjs"));
function createSsrBootstrapModule(rootSpecifier) {
  return `
import { renderComponent } from '@lwc/engine-server';
import Ctor, * as rootComponent from '${rootSpecifier}';

(async () => {
    // 1. setup page data
    const context = globalThis.getContext();
    let props = context.props;
    let markup;
    if (rootComponent.getPageData) {
        const data = await rootComponent.getPageData(context);
        props = data.props; // overwrite public props
        markup = data.markup;
    }

    // 2. render component
    const result = renderComponent('${(0, import_shared_utils.moduleSpecifierToKebabCase)(rootSpecifier)}', Ctor, props || {});

    // 3. relay results
    globalThis.resolver({ result, props, markup });
})()`;
}
var LwcSsrModuleProvider = class {
  constructor(providerConfig, {runtimeEnvironment: {lwrVersion}}) {
    this.name = "ssr-module-provider";
    this.version = lwrVersion;
  }
  async getModuleEntry({specifier}) {
    if (specifier.startsWith(import_identity.LWC_SSR_PREFIX)) {
      const virtualId = `<virtual>/${specifier}`;
      return {
        id: `${virtualId}|${this.version}`,
        virtual: true,
        entry: virtualId,
        specifier,
        version: this.version
      };
    }
  }
  async getModule({specifier, namespace, name}) {
    const moduleEntry = await this.getModuleEntry({specifier});
    if (!moduleEntry) {
      return;
    }
    const originalSource = createSsrBootstrapModule(specifier.replace(import_identity.LWC_SSR_PREFIX, ""));
    return {
      id: moduleEntry.id,
      namespace,
      name: name || specifier,
      version: this.version,
      specifier,
      moduleEntry,
      ownHash: (0, import_shared_utils.hashContent)(originalSource),
      originalSource,
      compiledSource: originalSource
    };
  }
};
var moduleProvider_default = LwcSsrModuleProvider;
